snap/info.go

```diff
@@ -950,6 +951,109 @@ func (s *Info) DesktopPrefix() string {
+	var desktopPlug *PlugInfo
+	for _, plug := range s.Plugs {
+		if plug.Interface == "desktop" {
+			desktopPlug = plug
+			break
+		}
+	}
+	if desktopPlug == nil {
+		return nil, nil
+	}
+
+	attrVal, exists := desktopPlug.Lookup("desktop-file-ids")
+	if !exists {
+		// desktop-file-ids attribute is optional
+		return nil, nil
+	}
+
+	// TODO: The internal errors below should never happen due to validation
+	// in the desktop interface. It would be a good candidate for telemetry
+	// error reporting.
+
+	// desktop-file-ids must be a list of strings
+	attrList, ok := attrVal.([]interface{})
+	if !ok {
+		return nil, errors.New(`internal error: "desktop-file-ids" must be a list of strings`)
+	}
+
+	desktopFileIDs := make([]string, 0, len(attrList))
+	for _, val := range attrList {
+		desktopFileID, ok := val.(string)
+		if !ok {
+			return nil, errors.New(`internal error: "desktop-file-ids" must be a list of strings`)
+		}
+		desktopFileIDs = append(desktopFileIDs, desktopFileID)
+	}
+	return desktopFileIDs, nil
+}
+
+// MangleDesktopFileName returns the file name prefixed with Info.DesktopPrefix()
+// unless its name (without the .desktop extension) is listed under the
+// desktop-file-ids desktop interface attribute.
+func (s *Info) MangleDesktopFileName(desktopFile string) (string, error) {
+	desktopFileIDs, err := s.DesktopFileIDs()
+	if err != nil {
+		return "", err
+	}
+	dir := filepath.Dir(desktopFile)
+	base := filepath.Base(desktopFile)
+	// Don't mangle desktop files if listed under desktop-file-ids attribute
+	// XXX: Do we want to fail if a desktop-file-ids entry doesn't
+	// have a corresponding file?
+	for _, desktopFileID := range desktopFileIDs {
+		if strings.TrimSuffix(base, ".desktop") == desktopFileID {
+			return filepath.Join(dir, base), nil
+		}
+	}
+	// FIXME: don't blindly use the snap desktop filename, mangle it
+	// but we can't just use the app name because a desktop file
+	// may call the same app with multiple parameters, e.g.
+	// --create-new, --open-existing etc
+	return filepath.Join(dir, fmt.Sprintf("%s_%s", s.DesktopPrefix(), base)), nil
+}
+
+type DesktopFilesFromMountOptions struct {
+	// Mangles found desktop files using Info.MangleDesktopFileName()
+	MangleFileNames bool
+}
+
+// DesktopFilesFromMount returns the desktop files found under <snap-mount>/meta/gui.
+func (s *Info) DesktopFilesFromMount(opts *DesktopFilesFromMountOptions) ([]string, error) {
+	if opts == nil {
+		opts = &DesktopFilesFromMountOptions{}
+	}
+
+	rootDir := filepath.Join(s.MountDir(), "meta", "gui")
+	if !osutil.IsDirectory(rootDir) {
+		return nil, nil
+	}
+	desktopFiles, err := filepath.Glob(filepath.Join(rootDir, "*.desktop"))
+	if err != nil {
+		return nil, fmt.Errorf("cannot get desktop files from %v: %s", rootDir, err)
+	}
+
+	if !opts.MangleFileNames {
+		return desktopFiles, nil
+	}
+
+	for i, df := range desktopFiles {
+		desktopFiles[i], err = s.MangleDesktopFileName(df)
```

Comment:
perhaps if do not export this function then it could take a list of declared desktop plug IDs? otherwise it'll end up calling the same DesktopFileIDs() for every desktop file we find

Reply:
I agree, I was hesitant about this, having it more readable vs more efficient, I went with the readable approach since it was exported.

Open to any suggestions on simplifying this while keeping the implementation consistent between the desktop wrappers and apparmor rules generation https://github.com/canonical/snapd/pull/14456#discussion_r1743223900